### ISSUE TITLE
[PT2] Don't graph break on loss.backward()

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#138217]
@vmoens mentioned this recently in a talk on TorchRL pain points but I didn't see an existing issue.

We can actually support loss.backward() in Dynamo. There are a couple of cases (ranging from easy to difficult):

1) If the only differentiable inputs to a torch.compile region are graph leaves, then we can handle all of this inside Dynamo: allow_in_graph some call to compute gradients in the graph and Dynamo can update the .grad field of the inputs. This essentially requires the full model to be compile able.
2) If any inputs are differentiable and not leaves: maybe we can do something with compiled autograd here. What if we could trigger compiled autograd and then inline the graph we get from that into the graph that Dynamo is producing?

**Comment by vmoens** [comment]
Thanks for this @zou3519 
I'd say that the vast majority of cases are going to fall in the first bucket (leaves only) so supporting that would already be a major upgrade!

**Comment by yf225** [comment]
Might be related: https://github.com/pytorch/pytorch/issues/117394

**Comment by xmfan** [comment]
(1) also only works when there's no backward hooks and no aot metadata speculation errors
